runtime/lib/ttnn/program.cpp

```diff
@@ -24,50 +25,109 @@
+         inputTensor.storage_type() == ::tt::tt_metal::StorageType::DEVICE);
+  const ::tt::target::MemorySpace targetMemorySpace =
+      op->out()->desc()->layout()->memory_desc()->memory_space();
+  switch (targetMemorySpace) {
+  case ::tt::target::MemorySpace::System:
+  case ::tt::target::MemorySpace::SystemMMIO: {
+    if (inputTensor.storage_type() == ::tt::tt_metal::StorageType::BORROWED) {
+      ::ttnn::Tensor hostTensor = inputTensor.to(::ttnn::ROW_MAJOR_LAYOUT);
+      hostTensor = ::ttnn::to_dtype(hostTensor, targetDataTypeTTNN);
+      ::ttnn::Tensor &outputTensor = *liveTensors.at(op->out()->global_id());
+      void *src = ::tt::tt_metal::get_raw_host_data_ptr(hostTensor);
+      void *dst = ::tt::tt_metal::get_raw_host_data_ptr(outputTensor);
+      std::uint32_t size = hostTensor.volume() * hostTensor.element_size();
+      std::memcpy(dst, src, size);
+    } else if (inputTensor.storage_type() ==
+               ::tt::tt_metal::StorageType::DEVICE) {
+      ::ttnn::Tensor hostTensor =
+          ::ttnn::typecast(inputTensor, targetDataTypeTTNN);
+      // Following the flow in core.py::to_torch - untilize on host
+      hostTensor = hostTensor.cpu().to(::ttnn::ROW_MAJOR_LAYOUT);
+      ::ttnn::Tensor &outputTensor = *liveTensors.at(op->out()->global_id());
+      void *src = ::tt::tt_metal::get_raw_host_data_ptr(hostTensor);
+      void *dst = ::tt::tt_metal::get_raw_host_data_ptr(outputTensor);
+      std::uint32_t size = hostTensor.volume() * hostTensor.element_size();
+      std::memcpy(dst, src, size);
     }
-    auto &outputTensor = *liveTensors.at(op->out()->global_id());
-    void *src = ::tt::tt_metal::get_raw_host_data_ptr(untilized);
-    void *dst = ::tt::tt_metal::get_raw_host_data_ptr(outputTensor);
-    std::uint32_t size = untilized.volume() * untilized.element_size();
-    std::memcpy(dst, src, size);
-    return;
-  }
-  bool isL1 = op->in0()->desc()->layout()->memory_desc()->memory_space() ==
-              ::tt::target::MemorySpace::DeviceL1;
-  const auto memoryConfig =
-      isL1 ? ::ttnn::L1_MEMORY_CONFIG : ::ttnn::DRAM_MEMORY_CONFIG;
-  auto &inputTensor = *liveTensors.at(op->in0()->global_id());
-  ::ttnn::Tensor tilized = ::tilize(inputTensor);
-  auto deviceTensor = ::ttnn::to_device(tilized, &device, memoryConfig);
-  tensorPool.push_back(deviceTensor);
-  // auto [iter, inserted] =
-  liveTensors.try_emplace(op->out()->global_id(), &tensorPool.back());
-  // assert(inserted && "Duplicate output tensor");
+    break;
+  }
+  case ::tt::target::MemorySpace::DeviceDRAM: {
+    ::tt::tt_metal::MemoryConfig memConfig = ::ttnn::DRAM_MEMORY_CONFIG;
+    // Host tensor, currently only supports borrowed storage
+    if (inputTensor.storage_type() == ::tt::tt_metal::StorageType::BORROWED) {
+      // moving to device first allows us to use device tilize
+      ::ttnn::Tensor deviceTensor =
+          ::ttnn::to_device(inputTensor, &device, memConfig);
+      deviceTensor = ::tilize(deviceTensor);
```

Comment:
Why do we tilize here?  I think we only need to if `MemoryDesc.tile_shape > {1, 1}`

Reply:
@nsmithtt Just updated this, but currently it seems that the tensor tile_shapes are hardcoded to 0x0 in the flatbuffer executable - my guess is that MLIR's builtin tensor type doesn't have knowledge about tile_shapes, and there are no compiler passes to update the tile_shapes as needed... This will result in the tilize code path to never be run, should I push a hack around this for now?